torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -143,66 +144,129 @@ class ProcessGroupNCCL : public ProcessGroup {
-  // Future interface.
+  // Future interface. FutureNCCL does not support NCCL_BLOCKING_WAIT flag
+  // or NCCL's barrier().
   //
-  // FutureNCCL has a reference to WorkNCCL and NCCL collective's outputs.
-  // Its value is NCCL collective's outputs.
+  // If created by WorkNCCL's getFuture API, FutureNCCL has a reference to
+  // WorkNCCL's cudaEvents, NCCL collective's outputs, and device index of
+  // outputs' device. Its value is NCCL collective's outputs. FutureNCCL
+  // only supports single-process single-device mode where the size of outputs
+  // is equal to 1.
+  //
+  // If created by FutureNCCL's then callback, its value becomes the value of
+  // callback() and its cudaEvent will record the NCCL stream that runs that
+  // callback.
   struct FutureNCCL : at::ivalue::Future {
    public:
     explicit FutureNCCL(
-        std::shared_ptr<ProcessGroupNCCL::WorkNCCL> work,
-        std::shared_ptr<std::vector<at::Tensor>> outputs)
+        at::IValue value,
+        c10::DeviceIndex deviceIndex,
+        std::shared_ptr<at::cuda::CUDAEvent> cudaEvent)
         : at::ivalue::Future(c10::ListType::create(c10::TensorType::get())),
-          work_(work),
-          outputs_(*outputs) {}
-
-    // Simply calls WorkNCCL's wait(). It will return after synchronizing
-    // the correct GPU streams to ensure we can have async CUDA execution
-    // and it does not wait for the entire operation to complete on GPU.
-    // If NCCL_BLOCKING_WAIT is enabled, in that case, it will wait for the
-    // entire operation to complete before returning.
+          value_(value),
+          deviceIndex_(deviceIndex),
+          cudaEvent_(cudaEvent) {}
+
+    // Gets current stream of device and synchronizes recorded streams
+    // with that. It will return after synchronizing the correct GPU streams to
+    // ensure we can have async CUDA execution and it does not wait for the
+    // entire operation to complete on GPU.
     void wait() override {
-      work_->wait();
+      auto currentStream = at::cuda::getCurrentCUDAStream(deviceIndex_);
+      (*cudaEvent_).block(currentStream);
     }
 
-    // FutureNCCL's value is NCCL collective's outputs and callbacks were
-    // invoked inline by addCallback(), so markCompleted is not needed.
+    // FutureNCCL's value is NCCL collective's outputs or callback() output with
+    // callbacks being invoked inline by addCallbackWithStream(), so
+    // markCompleted is not needed.
     void markCompleted(at::IValue /* unused */) override {
       C10_THROW_ERROR(Error, "FutureNCCL::markCompleted is not supported.");
     }
 
-    // Just returns NCCL collective's outputs after WorkNCCL's wait returns.
+    // Just returns FutureNCCL's value.
     at::IValue value() override {
-      work_->wait();
-      return outputs_;
+      return value_;
     }
 
     const at::IValue& constValue() override {
-      work_->wait();
-      return outputs_;
+      return value_;
```

Comment:
We should probably just call `wait()` in here to be safe if the user doesn't call `wait()` themselves. This should be fine in our callbacks, since `.value()` is called inside our newly created stream, so `wait()` will block on the newly created stream and not the default stream.